### ISSUE TITLE
Enum final checking false positive in complex case

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#11702]
This code generates a false positive, but weirdly I can only reproduce it if the base class is defined in a different module:

```py
from mod import E

class EE(E):  # Cannot inherit from final class "E"
    X = (0, "")
```

```py
# mod.py
from enum import Enum
from typing import Optional, Type

class E(int, Enum):
    def __init__(self, _: int, x: Optional[str] = None) -> None:
        self.x = x

    def __new__(cls: Type['E'], value: int, x: Optional[str] = None) -> 'E':
        enum = int.__new__(cls, value)
        enum._value_ = value
        enum.x = x
        return enum
```

The code works at runtime.

This started happening as a side effect of #11247.

Marking as high priority, since this is a regression. If it seems clear that this may only happen in pretty specialized circumstances, we can perhaps lower the priority.

cc @sobolevn

**Comment by sobolevn** [comment]
See https://github.com/python/mypy/pull/11579/files
Will it fix this case?

**Comment by sobolevn** [comment]
Looks like it fixes this issue indeed!
<img width="752" alt="Снимок экрана 2021-12-10 в 20 35 04" src="https://user-images.githubusercontent.com/4660275/145617146-f8859c11-5957-45a4-8617-356b901dc515.png">
<img width="752" alt="Снимок экрана 2021-12-10 в 20 35 20" src="https://user-images.githubusercontent.com/4660275/145617159-906c505d-752d-4c84-8f87-fcff716b8861.png">

**Comment by JukkaL** [comment]
Excellent! I can review the PR early next week.
